### PR TITLE
[sparse] add sparse grad wrapper

### PR DESCRIPTION
**Pull request by jakevdp**
This is a wrapper for `jax.grad` that correctly handles gradients with respect to a sparse matrix (i.e. does not differentiate the indices).

### CODE REVIEW

File: jax/experimental/sparse_ops.py [hunk 1]
```diff
@@ -51,11 +51,52 @@
...
 from jax.util import safe_zip
 from jax._src.lax.lax import ranges_like, remaining, _dot_general_batch_dim_nums
+from jax.api_util import _ensure_index, _ensure_index_tuple
 
 xb = xla_bridge
 xops = xla_client.ops
 Dtype = Any
 
+
+def grad(fun: Callable, argnums: Union[int, Sequence[int]] = 0, **kwargs) -> Callable:
```
Comment (shoyer) [inline jax/experimental/sparse_ops.py:61]: Maybe worth doing `value_and_grad`, and making `grad` a wrapper of that, as is the case in `api.py`? In my experience, it is pretty common to want to pull out the `loss` when doing gradient based optimization. (and it doesn't add any additional cost)

Comment (jakevdp) [reply jax/experimental/sparse_ops.py]: Good idea - Done!

File: jax/experimental/sparse/ad.py [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+from jax import core
+from jax import tree_util
+from jax.api_util import _ensure_index, _ensure_index_tuple
+from jax.util import safe_zip
+from jax._src.util import wraps
+from jax._src.traceback_util import api_boundary
+from .ops import BCOO
+
+
+def value_and_grad(fun: Callable, argnums: Union[int, Sequence[int]] = 0, **kwargs) -> Callable[..., Tuple[Any, Any]]:
```
Comment (mattjj) [inline jax/experimental/sparse/ad.py:29]: Could you try to wrap to 80 characters? It's not a strict rule but it helps some of our colleagues. This line seems super long. (I have a command in my editor which hard-wraps to 80 chars. Maybe yours can do that too!)

**Review by mattjj — APPROVED** [review]
LGTM!

File: jax/experimental/sparse/ad.py (hunk 2 above)
Comment (jakevdp) [reply jax/experimental/sparse/ad.py]: Done
